### PR TITLE
examples: Add LSRTM acoustic notebook

### PR DESCRIPTION
**Pull request by rabreucristo**
Hi folks,

Find here a LSRTM acoustic case notebook using Born operator.

Your suggestions and comments are more than welcome.

### CODE REVIEW

**Comment by speglich** [comment]
Hi @rabreucristo,

Cells 6 and 10 are missing the initial flag #NBVAL_IGNORE_OUTPUT. It would be interesting to start by correcting that.

**Comment by speglich** [comment]
@rabreucristo 

Please add the parameter '#NBVAL_SKIP' on the cells that are failing with `Timeout of 5 seconds exceeded waiting for output.`

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 1]
```diff
@@ -0,0 +1,11376 @@
...
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# First, position source centrally in all dimensions, then set depth\n",
+    "src_coordinates = np.empty((1, 2))\n",
+    "src_coordinates[0, :] = np.array(model.domain_size) * .5\n",
+    "src_coordinates[0, -1] = 20.  # Depth is 20m\n",
+    "\n",
+    "\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:236]: Too much skipped

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 2]
```diff
@@ -0,0 +1,11376 @@
...
+    "# Create image symbol and instantiate the previously defined imaging operator\n",
+    "image = Function(name='image', grid=model.grid)\n",
+    "op_imaging = ImagingOperator(model, image)\n",
+    "for i in range(nshots):\n",
+    "    print('Observed data %d out of %d' % (i+1, nshots))\n",
+    "    geometry.src_positions[0, :] = source_locations[i, :]\n",
+    "\n",
+    "    true_d , _, _ = solver.forward(vp=model.vp)\n",
+    "      \n",
+    "    print('Calculated data %d out of %d' % (i+1, nshots))\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:738]: Is all this verbosity with prints necessary?

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 3]
```diff
@@ -0,0 +1,11376 @@
...
+    "\n",
+    "# Define acquisition geometry: receivers\n",
+    "\n",
+    "# Initialize receivers for synthetic and imaging data\n",
+    "rec_coordinates = np.empty((nreceivers, 2))\n",
+    "rec_coordinates[:, 0] = np.linspace(0, model.domain_size[0], num=nreceivers)\n",
+    "rec_coordinates[:, 1] = 30.\n",
+    "\n",
+    "# Geometry\n",
+    "\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:245]: skipped line

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 4]
```diff
@@ -0,0 +1,11376 @@
...
+      "Operator `Kernel` ran in 0.01 s\n",
+      "Operator `Forward` ran in 0.07 s\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Observed data 6 out of 21\n"
+
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:11119]: Maybe add some comments?

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 5]
```diff
@@ -0,0 +1,11376 @@
...
+       "<Figure size 576x432 with 2 Axes>"
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:831]: The output of this cell is insanely giant, maybe reduce the verbosity of the operator? Or the steps?

**Review by speglich — COMMENTED** [review]
Hi @rabreucristo, some suggestions :)

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb (hunk 4 above)
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:11155]: Something less flashy?

**Comment by speglich** [comment]
Much better to understand / review now 😍

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+    "                         time_range=geometry.time_axis,\n",
+    "                         coordinates=geometry.rec_positions)\n",
+    "    \n",
+    "    dm =  image_up_dev.data/((model0.vp.data)**3) \n",
+    "\n",
+    "    dm_true =  (solver.model.vp.data**(-2) - model0.vp.data**(-2)) # Using devito's reference of virtual source\n",
+    "\n",
+    "    print('LSRTM Iteration',k+1)\n",
+    "    objective = 0.\n",
+    "    for i in range(nshots):\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:513]: Probably you could improve cell 11, where the LSRTM is performed. Maybe you could replace the interior loop shots of LSRTM by a function call. The `03_FWI` notebook use the same approach.  

As you declare the vast majority of the variables within that loop, I believe that is a simple change.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb (hunk 1 above)
Comment (speglich) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb]: @rabreucristo, up!

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb (hunk 6 above)
Comment (rabreucristo) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb]: Thank you @speglich . All the changes have been made. Let me know if it works for you.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+    "Now we are going to perform the LSRTM."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def lsrtm_gradient(nshots,mask,dm):\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:445]: Not all variables are being passed as parameters, many of them being global, so nshots can also be global.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+    "    \n",
+    "    residual_shot = Receiver(name='residual_shot', grid=model.grid, time_range=geometry.time_axis,\n",
+    "                   coordinates=geometry.rec_positions)\n",
+    "\n",
+    "    grad_full = Function(name='grad_full', grid=model.grid)\n",
+    "    \n",
+    "    residual = Receiver(name='residual', grid=model.grid, time_range=geometry.time_axis,\n",
+    "                   coordinates=geometry.rec_positions)\n",
+    "    \n",
+    "    d_obs = Receiver(name='d_obs', grid=model.grid,\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:455]: **Maybe** `time_range` could be in the same line that `grid` and `name`.

Comment (speglich) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:455]: These variables aren't inside of the inner loop, they must be here?

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+    "\n",
+    "    print('LSRTM Iteration',k+1)\n",
+    "        \n",
+    "    objective,residual_shot,grad_full = lsrtm_gradient(dm)\n",
+    "    \n",
+    "    residual_k[k] = residual_shot.data[:][:]\n",
+    "    \n",
+    "    history[k] = objective\n",
+    "    \n",
+    "    residual_num[:][:] = residual_k[k-1]-residual_k[k]\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:551]: Great Job @rabreucristo! Maybe you could remove this `[:][:]` to residual_num.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 10]
```diff
@@ -0,0 +1,746 @@
...
+    "    \n",
+    "    residual_num = np.zeros((geometry.nt,model.shape[1]))\n",
+    "    \n",
+    "    dm =  image_up_dev/((model0.vp.data)**3) \n",
+    "\n",
+    "    print('LSRTM Iteration',k+1)\n",
+    "        \n",
+    "    objective,residual_shot,grad_full = lsrtm_gradient(dm)\n",
+    "    \n",
+    "    residual_k[k] = residual_shot.data[:][:]\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:547]: Maybe you could remove this [:][:] from the residual_shot.data too

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 11]
```diff
@@ -526,39 +526,36 @@
...
     "    \n",
-    "    residual_k.data[k][:][:] = residual_shot.data[:][:]\n",
     "    history[k] = objective\n",
-    "    residual_num.data[:][:] = residual_k.data[k-1][:][:]-residual_k.data[k][:][:]\n",
-    "    history_norm[k] =norm(residual_num)**2/norm(residual_k)**2\n",
+    "    \n",
+    "    residual_num[:][:] = residual_k[k-1]-residual_k[k]\n",
```
Comment (speglich) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:551]: Maybe you could remove the `[:][:]` from residual_num.

**Comment by speglich** [comment]
Great Job @rabreucristo! Thank you for your patience and attention.

**Comment by speglich** [comment]
@nogueirapeterson I believe that at least on the computing side we have a more mature PR, could you take a look at the  geophysics area?

**Comment by rabreucristo** [comment]
Thank you @speglich. The changings have been made! @nogueirapeterson from now on I think we can work on your feedbacks! I believe will be a long journey! :)

**Comment by nogueirapeterson** [comment]
> @nogueirapeterson I believe that at least on the computing side we have a more mature PR, could you take a look at the geophysics area?

Yes, João. I will do that soon.

**Comment by nogueirapeterson** [comment]
@rabreucristo , congratulations on your contribution. This is a really interesting tutorial!

After @speglich review with respect to Devito's programming, I left below some doubts, tips, and suggestions to contribute to the improvement of his tutorial, with respect to the geophysical part:

I don't think you should use the term “Born reverse modeling”. Perhaps “adjoint modeling” is more appropriate.

Are you considering the initial reflectivity model as the conventional RTM or as the first iteration of the gradient? If it is the conventional RTM, I think it is more appropriate for you to apply the conventional imaging condition:` Im = Im + us * ur` (check the signal).
But if you are using the gradient, check the image condition. Because I saw that you scale by` 1 / v ** 3.`

I think that in your inversion scheme, you can normalize the amplitude between the calculated data and the observed data. In this way, you will produce a more consistent residue data and, thus, will help to obtain a better convergence and consequently a better-estimated reflectivity model.

You can also include preconditioning in your inversion. You can calculate the sum of the square of the wavefield for all time steps and all the shots (an approximation of the Hessian diagonal) and then divide by the gradient. This should help a lot to improve the quality of the inverted reflectivity model.

Do you know why this artifact happens? Maybe @mloubout  knows the origin of this artifact.
I think that, instead of applying a tapered mask, you can update the current model only in the region below the water depth and preserve samples that correspond to the same water depth as the real model. So, in this case, you can use a different velocity model with three layers, etc... to make your experiment more interesting.

I think the step length can be calculated using a method more consistent. For example, in the first iteration, you can use the analytical step length derived from the steepest descent method. As for the following step length, you can use the Barzilai Borwein (BB) method, which is a simple method and works well.


Did you calculate `dm` as `Im / vp ** 3 `and `dm_true = 1 / vp ** 2 - 1 / vp ** 2`? Is this consistent? Could you explain me?

Why are you using the normalized objective function? I believe that if you change your objective function, you will consequently have to change your adjoint source.

Finally, I think you could better explore the steps of the LSRTM. For example, I think it would be interesting to plot the wavefields `u0` (background wavefield) and `du` (scattered wavefield) and their corresponding seismic data for some iterations. I think it will give more value to your tutorial, in addition to being more didactic and easier to interpret the results.

Once again, thank you very much for your contribution.
I hope I helped in the comments.

**Review by nogueirapeterson — COMMENTED** [review]
Hi @rabreucristo,

I left some comments on "conversations".

**Comment by rabreucristo** [comment]
Hi @nogueirapeterson, thanks for your comments. I will answer your questions in the order will appear.
1) I changed do adjoint modelling
2&3) The RTM I am using here is not the conventional RTM, I am always migrating the residue instead of migrating the observed data. Actually The image is the gradient. For the first iteration my gradient is zero so the residue is only the observed data. Doing that I am kind of doing a normalization because the amplitudes of my gradient will produce a calculated data with amplitude range of the observed data.
4)Nice, I can include this normalization!
5) I have suffered a lot to remove this artefact but I could not understand @mloubout the origin of that( even though Mathias has exlpained it to me) The mask was a quick and dirty solution to stabilize the problem. Yeah, I can perform a three-layer-model later once I have everything all set.
6) Alright, I will change the step length!
7) This is a good question! Theory-wise I am not sure, I think @mloubout & @jkwashbourne  could help me on that, but I can say they are consistent otherwise the inversion would not converge.
8) Normalized objective function is just for plotting purpose.
9)I will plot the wavefields, and also the true reflectivity, inverted one and initial one.

Well, I have a lot of work to do :) Once I have an update I will comit again! Meanwhile, feel free to comment anything and thank you for your comments!

**Comment by FabioLuporini** [comment]
Hey @rabreucristo @nogueirapeterson @speglich -- any news on this front?

**Comment by rabreucristo** [comment]
> Hey @rabreucristo @nogueirapeterson @speglich -- any news on this front?

Hi, @FabioLuporini. I have already done all the changes suggested by @nogueirapeterson but I have 4 failings when I try to commit the notebook. @speglich will help me on that.

**Comment by rabreucristo** [comment]
Hi @FabioLuporini, could you help me with this error? I really don't know what is happening about the environment.

**Comment by nogueirapeterson** [comment]
@rabreucristo  good job!
The tutorial is becoming very nice!
I think you missed some points:
 - The normalization between seismic data
 - The Plotting of the wavefields
I didn't quite understand your intention with zooming in on the objective function.

**Comment by nogueirapeterson** [comment]
> Hey @rabreucristo @nogueirapeterson @speglich -- any news on this front?

The tutorial improved after the major review.
Now, I just left some points for a minor review.

**Comment by rabreucristo** [comment]
> > Hey @rabreucristo @nogueirapeterson @speglich -- any news on this front?
> 
> The tutorial improved after the major review.
> Now, I just left some points for a minor review.

As discussed, I have done all the changings! Hopefully it can be successfully pushed and wait for the other reviewers.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb [hunk 12]
```diff
@@ -0,0 +1,718 @@
+{
```
Comment (mloubout) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:1]: <p>Can you double check that section. Some part of the text seem to be missing or werirdly ordered. For example at the begging you have</p><pre class="ql-syntax" spellcheck="false">into the constant-density acoustic wave equation, considering&nbsp;

</pre><p>that is missing the beginning of the sentence.</p><p> </p><ul><li><code>q</code>  is usually for the source, use <code>v</code>  or <code>u_r</code>  for the adjoint wavefield</li><li>Developing =&gt; expanding</li><li>	"Combining the Born's approximation into equation (2). That's not actually correct. You are substituting your equation 3a into 2 to cancel out the background wavefield to get equation 3b.</li><li>Concerning eq 5, put ref to tutorial that derives it.</li><li>For equation (6) this is not true/incomplete. Least square migration is a least-square problem <code>min_m ||J*dm - \delta d||</code>  which gradient is equation 5 and <code>J</code>  is the Jacobian or Born modeling operator</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1574/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (mloubout) [inline examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:1]: <p>I wouldn't put this one, desn't really mean anything.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/devitocodes/devito/pull/1574/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (rabreucristo) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:1]: Thank you, @mloubout. I will fix it.

Comment (mloubout) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:1]: <p>Oh right indeed, missread it. Would it be possible to rename the variables into more meaningful and readable names, such as <code>rtm</code>,  <code>lsrtm</code>,  <code>dm</code>  <span class="ql-cursor">﻿</span></p>

**Comment by mloubout** [comment]
Also, the `src_illum` term isn't described or explained.

File: examples/seismic/tutorials/13_LSRTM_acoustic.ipynb (hunk 12 above)
Comment (rabreucristo) [reply examples/seismic/tutorials/13_LSRTM_acoustic.ipynb:1]: Sure! I will do it!

**Comment by rabreucristo** [comment]
Hi @mloubout. I have done all the suggestions you mentioned. 

About the reference of equation (5), here is my reference http://dx.doi.org/10.22564/rbgf.v34i3.831. 

About this comment "For equation (6) this is not true/incomplete. Least square migration is a least-square problem min_m ||J*dm - \delta d|| which gradient is equation 5 and J is the Jacobian or Born modeling operator" This equation is just for the migration, the LSRTM itself is described in the next equation, referring to an iterative method.

Let me know any additional comments,

Thank you.

**Comment by mloubout** [comment]
>This equation is just for the migration, the LSRTM itself is described in the next equation, referring to an iterative method.

This next equation is not "LSRTM", it is the gradient descent update used to solve the least square problem. Least-square stands for the least square minimization problem you are solving `min_{dm} ||J*dm - d||` instead of just computing an image. The "iterative method" is the optimization algorithm, the "least square" is the problem you are solving, and the distinction has to be made as there is a variety of options for both the algorithm and the minimized objective that can lead to a completely different formulation.

**Comment by rabreucristo** [comment]
> > This equation is just for the migration, the LSRTM itself is described in the next equation, referring to an iterative method.
> 
> This next equation is not "LSRTM", it is the gradient descent update used to solve the least square problem. Least-square stands for the least square minimization problem you are solving `min_{dm} ||J*dm - d||` instead of just computing an image. The "iterative method" is the optimization algorithm, the "least square" is the problem you are solving, and the distinction has to be made as there is a variety of options for both the algorithm and the minimized objective that can lead to a completely different formulation.

got you! I will fix that.

**Comment by rabreucristo** [comment]
@mloubout I have fixed the notebook. Please, check it out and let me know! Thank you!

**Review by mloubout — APPROVED** [review]
Could use another pass for the description and  extra details, but overall looks fine and is a nice addition.

**Pull request merged**